api/client/service.proto

```diff
@@ -34,10 +36,59 @@ message ExecutionResponse {
   google.rpc.Status error_detail = 7;
 }
 
+// Used to request a RequestStreamResponse.
+message RequestStreamRequest {
+  // Amount of RequestStreamResponses that should be yielded.
+  // Used to implement basic flow control by the client. Once AsyncClientImpl gets
+  // a way to apply backpressure this could probably be dropped.
+  uint64 amount = 1;
+}
+
+message RequestStreamResponse {
+  // Specifies what the requests look like. Will be merged with the client's configuration.
+  RequestSpecifier request_specifier = 1;
+  // Response-level expectations associated to the above request specification.
+  Expectations expectations = 2;
+}
+
+message RequestSpecifier {
+  // method/path/authority/content_length can all be expressed in request headers.
+  // so technically they are redundant. In case of ambiguity, the explicit fields will prevail.
+  // Note that as of today, the target upstream/cluster that will be used to replay traffic to
+  // fixes the destination address/port.
+  // Request method
+  google.protobuf.StringValue method = 1;
+  // Request path
+  google.protobuf.StringValue path = 2;
+  // Request authority
+  google.protobuf.StringValue authority = 3;
+  // Request content length. The client will transfer the number of bytes specified here for the
+  // request body.
+  google.protobuf.UInt32Value content_length = 4;
+  // Request header replacements. Any existing header(s) with the same name will be removed
+  // before setting.
+  envoy.api.v2.core.HeaderMap headers = 5;
+  // TODO(oschaaf): nice to have
+  // google.protobuf.StringValue sni_hostname = 10;
+}
+
+// Response-level expectations. If set, any unmet expectations will result in stats counter
+// increases in the future.
+message Expectations {
+  // Expected response status code
+  google.protobuf.UInt32Value response_code = 1;
+  // Expected response body content length
+  google.protobuf.UInt32Value content_length = 2;
+}
+
 service NighthawkService {
   //  Accepts a stream of ExecutionRequests, which can be used to initiate, update, or cancel a
   //  test. Returns a stream of ExecutionResponse objects which will contain results for all
   //  Nighthawk runs.
   rpc ExecutionStream(stream ExecutionRequest) returns (stream ExecutionResponse) {
   }
+
+  // Obtains a stream request-level specifications plus expectations.
+  rpc RequestStream(stream RequestStreamRequest) returns (stream RequestStreamResponse) {
```

Comment:
To start a discussion - one point we were wondering about is whether this new RPC method would work better under a separate (new) service definition.

If I understand the architecture correctly, Nighthawk is meant to connect to some remote service to request the stream of request specifiers for the test run. Let's call that remote service the "traffic source". If this is true, we are really dealing with two binaries, Nighthawk itself and the traffic source.

Nighthawk should implement the ExecutionStream while the traffic source should implement the RequestStream. The traffic source shouldn't need to implement the ExecutionStream rpc, but any implementations of it will have to implement it to satisfy the service definition.

If I do have correct understanding - would it make sense to start by defining a separate service (e.g. TrafficSourceService) for the traffic source binary to implement? Alternatively please correct my understanding.

Reply:
A separate service absolutely makes sense to me. A separate binary as well.

Reply:
Can we make the corresponding change (splitting this out into a standalone service definition) before we move forward with this PR?

Reply:
Never mind moving the service definition into `api/service/`, that's probably a breaking change. 
Split into two separate service definitions in 8065182ef35c1eda851107750d7a9f77941251da

Reply:
Moving this new service out of the client subdirectory actually isn't a bad idea, because it doesn't logically fit into what is understood as the "client".

Can you help me understand why would it be a breaking change? The new service comes with a set of exclusively new messages that don't seem to have any dependencies on other types defined in the client protos. Am I missing something?

If we do end up moving it, I would recommend `api/request_source` since `api/service` feels bit too generic.

Reply:
To be honest, I'm not sure what I was thinking, this wouldn't be a breaking change. Done in 1dd144a99ea5af9025af0d705d81e660659d68e7